File: sdk/python/flatmachines/flatmachines/adapters/claude_code_sessions.py
"""Session holdback pattern for Claude Code CLI adapter.

Maintains a frozen "holdback" session that forks produce cache-warm
children from.  The holdback is never advanced — only forks diverge.

Cache mechanics:
  - Claude Code stores conversation history locally, sends to API on resume
  - API prompt cache is prefix-based with 1-hour TTL (Claude Max plan)
  - Fork sends the same prefix as the holdback → API cache hit
  - Periodic warm() keeps the prefix in API cache without advancing holdback

Cache writes are available immediately after seed returns — no warm
step is required before parallel fan-out.  warm() exists only to
reset the 1-hour TTL during long idle periods.

Usage:
    holdback = SessionHoldback(executor)
    await holdback.seed("Read the codebase and understand the architecture.")
    # Cache is ready — fan out immediately

    # Fan out — each gets full context, hits cache
    results = await holdback.fork_n([
        "Implement the auth module",
        "Implement the database layer",
        "Write the test suite",
    ])

    # Keep cache warm while idle (optional, within 1hr TTL)
    await holdback.warm()
"""

from __future__ import annotations

import asyncio
import logging
import uuid
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .claude_code import ClaudeCodeExecutor
from ..agents import AgentResult

logger = logging.getLogger(__name__)


@dataclass
class ForkResult:
    """Result from a forked session."""
    session_id: str
    task: str
    result: AgentResult
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0


@dataclass
class SessionHoldback:
    """Manages a frozen session for cache-warm fan-out.

    The holdback session is seeded once, then only forked from.
    Plain --resume is never used on the holdback — every operation
    after seed uses --fork-session to get a new session ID while
    preserving the shared prefix for API cache hits.

    Args:
        executor: ClaudeCodeExecutor instance (already configured)
        session_id: Optional pre-existing session ID to adopt as holdback
    """

    executor: ClaudeCodeExecutor
    session_id: Optional[str] = None
    _seeded: bool = field(default=False, init=False)
    _fork_count: int = field(default=0, init=False)
    _total_cost: float = field(default=0.0, init=False)

    async def seed(
        self,
        task: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> AgentResult:
        """Create the holdback session with initial content.

        This is the only call that uses --session-id (new session).
        All subsequent operations use --fork-session.

        The API prefix cache is available immediately after this returns.
        No warm step is needed before parallel fan-out.

        Args:
            task: Initial prompt to seed the session with
            context: Optional context for working_dir resolution

        Returns:
            AgentResult from the seed invocation
        """
        if self.session_id is None:
            self.session_id = str(uuid.uuid4())

        result = await self.executor._invoke_once(
            task=task,
            session_id=self.session_id,
            resume=False,
            context=context,
        )

        self._seeded = True
        self._accumulate_cost(result)

        logger.info(
            "Holdback seeded: session=%s cache_write=%s",
            self.session_id,
            (result.usage or {}).get("cache_write_tokens", 0),
        )

        return result

    async def adopt(self, session_id: str) -> None:
        """Adopt an existing session as the holdback.

        Use this when you have a session from a prior machine execution
        and want to fork from it without re-seeding.

        If the session has been idle for close to 1 hour, call warm()
        before fork_n() to reset the cache TTL.

        Args:
            session_id: Existing Claude Code session ID
        """
        self.session_id = session_id
        self._seeded = True
        logger.info("Holdback adopted: session=%s", session_id)

    async def fork(
        self,
        task: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> ForkResult:
        """Fork a new session from the holdback and execute a task.

        The fork gets a new session ID but sees the full holdback
        conversation.  The holdback is not advanced.

        Args:
            task: Prompt for the forked session
            context: Optional context for working_dir resolution

        Returns:
            ForkResult with the new session ID and result
        """
        if not self._seeded:
            raise RuntimeError("Holdback not seeded — call seed() or adopt() first")

        result = await self.executor._invoke_once(
            task=task,
            session_id=self.session_id,
            resume=True,
            context=context,
            fork_session=True,
        )

        self._fork_count += 1
        self._accumulate_cost(result)

        usage = result.usage or {}
        fork_session = (result.metadata or {}).get("session_id", "?")

        fr = ForkResult(
            session_id=fork_session,
            task=task,
            result=result,
            cache_read_tokens=usage.get("cache_read_tokens", 0),
            cache_write_tokens=usage.get("cache_write_tokens", 0),
        )

        logger.info(
            "Holdback fork: parent=%s child=%s cache_read=%s cache_write=%s",
            self.session_id, fork_session,
            fr.cache_read_tokens, fr.cache_write_tokens,
        )
        return fr

    async def fork_n(
        self,
        tasks: List[str],
        context: Optional[Dict[str, Any]] = None,
        max_concurrent: Optional[int] = None,
    ) -> List[ForkResult]:
        """Fork N sessions in parallel, each with a different task.

        All forks share the holdback prefix and hit API cache.

        Args:
            tasks: List of prompts, one per fork
            context: Optional context for working_dir resolution
            max_concurrent: Max concurrent forks (None = all at once)

        Returns:
            List of ForkResult in same order as tasks
        """
        if not self._seeded:
            raise RuntimeError("Holdback not seeded — call seed() or adopt() first")

        if max_concurrent is None or max_concurrent <= 0:
            max_concurrent = len(tasks)

        semaphore = asyncio.Semaphore(max_concurrent)

        async def _limited_fork(task: str) -> ForkResult:
            async with semaphore:
                return await self.fork(task, context)

        results = await asyncio.gather(
            *[_limited_fork(t) for t in tasks],
            return_exceptions=True,
        )

        # Convert exceptions to ForkResult with error
        final: List[ForkResult] = []
        for i, r in enumerate(results):
            if isinstance(r, Exception):
                final.append(ForkResult(
                    session_id="",
                    task=tasks[i],
                    result=AgentResult(
                        error={
                            "code": "server_error",
                            "type": type(r).__name__,
                            "message": str(r),
                            "retryable": False,
                        },
                        finish_reason="error",
                    ),
                ))
            else:
                final.append(r)

        return final

    async def warm(
        self,
        context: Optional[Dict[str, Any]] = None,
    ) -> AgentResult:
        """Send a minimal request to reset the API prefix cache TTL.

        Only needed if the holdback has been idle for close to 1 hour.
        Uses --fork-session so the holdback is not advanced.
        The fork is discarded — its only purpose is to touch the API
        with the holdback's prefix so the 1-hour cache TTL resets.

        Args:
            context: Optional context for working_dir resolution

        Returns:
            AgentResult from the warm invocation
        """
        if not self._seeded:
            raise RuntimeError("Holdback not seeded — call seed() or adopt() first")

        result = await self.executor._invoke_once(
            task="test",
            session_id=self.session_id,
            resume=True,
            context=context,
            fork_session=True,
        )

        self._accumulate_cost(result)

        usage = result.usage or {}
        logger.info(
            "Holdback warm: session=%s cache_read=%s cost=%s",
            self.session_id,
            usage.get("cache_read_tokens", 0),
            result.cost,
        )
        return result

    @property
    def stats(self) -> Dict[str, Any]:
        """Current holdback statistics."""
        return {
            "session_id": self.session_id,
            "seeded": self._seeded,
            "fork_count": self._fork_count,
            "total_cost": self._total_cost,
        }

    # -- Internal -----------------------------------------------------------

    def _accumulate_cost(self, result: AgentResult) -> None:
        if result.cost is not None:
            if isinstance(result.cost, (int, float)):
                self._total_cost += float(result.cost)
            elif isinstance(result.cost, dict):
                self._total_cost += float(result.cost.get("total", 0))

File: sdk/js/packages/flatmachines/src/adapters/claude_code_sessions.ts
/**
 * Session holdback pattern for Claude Code CLI adapter.
 */

import { randomUUID } from 'node:crypto';
import { ClaudeCodeExecutor } from './claude_code_adapter';

export class ForkResult {
  session_id: string;
  task: string;
  result: any;
  cache_read_tokens: number;
  cache_write_tokens: number;

  constructor(opts: { session_id: string; task: string; result: any; cache_read_tokens: number; cache_write_tokens: number }) {
    this.session_id = opts.session_id;
    this.task = opts.task;
    this.result = opts.result;
    this.cache_read_tokens = opts.cache_read_tokens;
    this.cache_write_tokens = opts.cache_write_tokens;
  }
}

export class SessionHoldback {
  public executor: ClaudeCodeExecutor;
  public session_id: string | null;
  public _seeded: boolean = false;
  public _fork_count: number = 0;
  public _total_cost: number = 0;

  constructor(executor: ClaudeCodeExecutor, sessionId?: string) {
    this.executor = executor;
    this.session_id = sessionId ?? null;
  }

  async seed(task: string, context?: Record<string, any>): Promise<any> {
    if (this.session_id == null) {
      this.session_id = randomUUID();
    }

    const invoker = (this.executor as any)._invoke_once ?? (this.executor as any).invokeOnce;
    if (!invoker) throw new Error('Executor missing _invoke_once method');

    let result: any;
    try {
      result = await invoker.call(this.executor, {
        task,
        session_id: this.session_id,
        resume: false,
        context,
      });
    } catch (e) {
      this._seeded = true;
      throw e;
    }

    this._seeded = true;
    this._accumulate_cost(result);
    return result;
  }

  async adopt(sessionId: string): Promise<void> {
    this.session_id = sessionId;
    this._seeded = true;
  }

  async fork(task: string, context?: Record<string, any>): Promise<ForkResult> {
    if (!this._seeded) {
      throw new Error('Holdback not seeded — call seed() or adopt() first');
    }

    const invoker = (this.executor as any)._invoke_once ?? (this.executor as any).invokeOnce;
    if (!invoker) throw new Error('Executor missing _invoke_once method');

    const result = await invoker.call(this.executor, {
      task,
      session_id: this.session_id,
      resume: true,
      context,
      fork_session: true,
    });

    this._fork_count += 1;
    this._accumulate_cost(result);

    const usage = result?.usage ?? {};
    const forkSession = result?.metadata?.session_id ?? result?.session_id ?? '?';

    return new ForkResult({
      session_id: forkSession,
      task,
      result,
      cache_read_tokens: usage.cache_read_tokens ?? 0,
      cache_write_tokens: usage.cache_write_tokens ?? 0,
    });
  }

  async fork_n(
    tasks: string[],
    context?: Record<string, any>,
    maxConcurrent?: number,
  ): Promise<ForkResult[]> {
    if (!this._seeded) {
      throw new Error('Holdback not seeded — call seed() or adopt() first');
    }

    const concurrency = maxConcurrent && maxConcurrent > 0 ? maxConcurrent : tasks.length;
    const results: ForkResult[] = new Array(tasks.length);
    let active = 0;
    let nextIndex = 0;
    const queue: Array<() => void> = [];

    const forkOne = async (index: number): Promise<void> => {
      try {
        results[index] = await this.fork(tasks[index]!, context);
      } catch (e: any) {
        results[index] = new ForkResult({
          session_id: '',
          task: tasks[index]!,
          result: {
            error: {
              code: 'server_error',
              type: e?.name ?? 'Error',
              message: e?.message ?? String(e),
              retryable: false,
            },
            finish_reason: 'error',
          },
          cache_read_tokens: 0,
          cache_write_tokens: 0,
        });
      }
    };

    // Simple semaphore-based concurrent execution
    const promises: Promise<void>[] = [];
    for (let i = 0; i < tasks.length; i++) {
      if (concurrency >= tasks.length) {
        promises.push(forkOne(i));
      } else {
        // Use a semaphore pattern
        promises.push((async () => {
          while (active >= concurrency) {
            await new Promise<void>(resolve => queue.push(resolve));
          }
          active++;
          try {
            await forkOne(i);
          } finally {
            active--;
            if (queue.length) queue.shift()!();
          }
        })());
      }
    }

    await Promise.all(promises);
    return results;
  }

  async warm(context?: Record<string, any>): Promise<any> {
    if (!this._seeded) {
      throw new Error('Holdback not seeded — call seed() or adopt() first');
    }

    const invoker = (this.executor as any)._invoke_once ?? (this.executor as any).invokeOnce;
    if (!invoker) throw new Error('Executor missing _invoke_once method');

    const result = await invoker.call(this.executor, {
      task: 'test',
      session_id: this.session_id,
      resume: true,
      context,
      fork_session: true,
    });

    this._accumulate_cost(result);
    return result;
  }

  get stats(): Record<string, any> {
    return {
      session_id: this.session_id,
      seeded: this._seeded,
      fork_count: this._fork_count,
      total_cost: this._total_cost,
    };
  }

  private _accumulate_cost(result: any): void {
    if (result?.cost != null) {
      if (typeof result.cost === 'number') {
        this._total_cost += result.cost;
      } else if (typeof result.cost === 'object' && result.cost.total != null) {
        this._total_cost += Number(result.cost.total);
      }
    }
  }
}